### PR TITLE
More Typescript migrations - switch to Typedoc

### PR DESCRIPTION
**Pull request by joepio**
Related issue: #355

- Converted some of the most fundamental classes to typescript, including Node, Literal, BlankNode, NamedNode, Collection, Statement.
- Introduced a `.types` file for shared types.
- Included a _temporary_ `types-temp.ts` file in project root as a reference file for documentation and keeping track of the ts migration process. 
- The `.isVar` method is set to boolean values, instead of 0 or 1. This seemed reasonable, as it's only used for boolean type checks, and the existing types already define it as a `boolean` value.
- JSDoc is switched with Typedoc. Combined with typescript, it makes the documentation far more complete.
- Sometimes I had to make assumptions on date types, e.g. that a user will never create a Statement containing a Collection. These assumptions used to be implicit, but typescript forces us to make them explicit. This happens when you see [type assertions](https://basarat.gitbooks.io/typescript/docs/types/type-assertion.html) `(value as type)`.
- Literals can apparently be `null` or `undefined`, when nodes are created using the `.fromValue` method. This happens in the Statement constructor. See #362.

### CODE REVIEW

File: src/node.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ * @method fromValue
+ * @param value - Any native Javascript value
+ */
+Node.fromValue = function (value: ValueType): Node | Literal | undefined | null | Collection {
+  if (typeof value === 'undefined' || value === null) {
+    return value
+  }
+  const isNode = Object.prototype.hasOwnProperty.call(value, 'termType')
+  if (isNode) {  // a Node subclass or a Collection
+    // @ts-ignore
```
Comment (Vinnl) [inline src/node.ts]: I'd preferably avoid adding `@ts-ignore` if possible, but if not, please at least add a comment indicating why it is required here, e.g.

    // @ts-ignore: Cannot cast `value` to the the required type because...

I did not add a comment to every `@ts-ignore`, as I'm sure you can do a Find in all Files ;)

File: src/node.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    return value
+  }
+  if (Array.isArray(value)) {
+    return new Collection(value)
+  }
+  return Literal.fromValue(value)
+}
+
+import Namespace from './namespace'
```
Comment (Vinnl) [inline src/node.ts]: I think usually it's good practice to put all `import` statements at the top of the file, so that it's clear when there are e.g. duplicate entries?

File: src/node.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  }
+  // Node remains Node
+  // @ts-ignore
+  if (!term.datatype) return term // Objects remain objects
+  const literalTerm = term as Literal
+  // if (!Object.prototype.hasOwnProperty.call(term, 'dataType')) return term // Objects remain objects
+  if (literalTerm.datatype.sameTerm(ns.xsd('boolean'))) {
+    return literalTerm.value === '1'
+  }
+  console.log("4", term)
```
Comment (Vinnl) [inline src/node.ts]: Forgot to remove this?
```suggestion
```

File: src/node.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  }
+  if (
+    literalTerm.datatype.sameTerm(ns.xsd('integer')) ||
+    literalTerm.datatype.sameTerm(ns.xsd('float')) ||
+    literalTerm.datatype.sameTerm(ns.xsd('decimal'))
+  ) {
+    let z = Number(literalTerm.value)
+    return Number(literalTerm.value)
+  }
+  console.log("7", term)
```
Comment (Vinnl) [inline src/node.ts]: Forgot to remove this?
```suggestion
```

File: src/store.js [hunk 5]
```diff
@@ -20,7 +20,7 @@ import { ArrayIndexOf } from './util'
 import Formula from './formula'
 import { RDFArrayRemove } from './util'
 import Statement from './statement'
-import Node from './node'
```
Comment (Vinnl) [inline src/store.js]: Hmm, this might just be my unfamiliarity with the codebase, but is this change intended/valid? Likewise for the same change in `variable.js`, `formula.js`, etc.

File: src/named-node.ts [hunk 6]
```diff
@@ -1,54 +1,58 @@
...
+   * @param iri The IRI for this node
    */
-  constructor (iri) {
+  constructor (iri: NamedNode | string) {
     super()
     this.termType = NamedNode.termType
 
-    if (iri && iri.termType === NamedNode.termType) {  // param is a named node
-      iri = iri.value
+    if (iri && (iri as NamedNode).termType === NamedNode.termType) {  // param is a named node
```
Comment (Vinnl) [inline src/named-node.ts]: Tip: create a [user-defined type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) (e.g. `isNamedNode()`), then there's no need to do the cast inside the `if` block. (Or even after it, I guess, since `string` is the complementary type.)

File: src/named-node.ts [hunk 7]
```diff
@@ -1,54 +1,58 @@
...
      return new NamedNode(str.slice(0, p + 1))
    }
-   /**
-    * Returns an NN for the whole web site, ending in slash.
-    * Contrast with the "origin" which does NOT have a trailing slash
-    */
-  site () {
+
+  /**
+   * Returns an NamedNOde for the whole web site, ending in slash.
```
Comment (Vinnl) [inline src/named-node.ts]: Heh, +1 for expanding the acronym :)

File: src/literal.ts [hunk 8]
```diff
@@ -1,13 +1,36 @@
...
+   * The language for the literal
+   */
+  lang!: string
+
+  /**
+   * The literal's datatype as a named node
+   */
+  datatype!: NamedNode
+
+  termType: string
```
Comment (Vinnl) [inline src/literal.ts]: Given that this can only be one string, this might be better?
```suggestion
  termType: 'Literal'
```

File: src/literal.ts [hunk 9]
```diff
@@ -108,25 +137,26 @@ export default class Literal extends Node {
...
+
   /**
-   * @method fromValue
-   * @param value
-   * @return {Literal}
+   * Builds a literal node from an input value
+   * @param value The input value
    */
-  static fromValue (value) {
+  static fromValue(value: ValueType): Literal | Node | undefined | null {
```
Comment (Vinnl) [inline src/literal.ts]: The if statements in here could also be well-served by some user-defined type guards, I think.

File: src/literal.ts [hunk 10]
```diff
@@ -1,13 +1,36 @@
...
+  /**
+   * Initializes this literal
+   * @param value The literal's lexical value
+   * @param language The language for the literal
+   * @param datatype The literal's datatype as a named node
+   */
+  constructor (value: string, language?: string | null, datatype?: NamedNode) {
     super()
-    this.termType = Literal.termType
+    this.termType = 'Literal'
```
Comment (Vinnl) [inline src/literal.ts]: Haha, how on earth did this work before? :P

File: src/store.js (hunk 5 above)
Comment (joepio) [reply src/store.js]: Note sure about the implications, I'll revert them

**Comment by joepio** [comment]
Thanks for all the feedback @Vinnl, I've made the changes you requested (except for the type guard in literal.ts, that changes the behavior and was a little tricky).

**Comment by joepio** [comment]
@Vinnl I've added these type guards in [a new branch](https://github.com/ontola/rdflib.js/tree/typescript-3). Shall I update this PR to include the changes from that branch, or shall I submit it as a separate one?

**Comment by Vinnl** [comment]
@joepio Generally, separate PRs for separate functionality is easier to review for me, so I'd prefer a new one, marked as dependent on this one. I see [more changes than just type guards](https://github.com/ontola/rdflib.js/compare/typescript-2...typescript-3) though :)

**Comment by joepio** [comment]
> @joepio I see [more changes than just type guards](https://github.com/ontola/rdflib.js/compare/typescript-2...typescript-3) though :)

Very true - the biggest one is RDFJS Taskforce compatibility.

**Comment by joepio** [comment]
Yesterday discussed to first merge #358. I'll rebase my typescript fork (which includes the changes from this PR + a lot more) on master after that, and then will issue a new PR.

**Comment by Vinnl** [comment]
@joepio #358 has now been merged, so feel free to rebase.

**Comment by joepio** [comment]
Replaced by #373
